### PR TITLE
[Filter] reload shared model @open sesame 2/28 00:30

### PR DESCRIPTION
**Pull request by kimjh12**
reload shared interpreter for filter objects sharing same key

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: gst/nnstreamer/tensor_filter/tensor_filter_common.c [hunk 1]
```diff
@@ -2978,3 +2978,28 @@ nnstreamer_filter_get_shared_model_representation (const char *key)
...
+  GstTensorFilterSharedModelRepresenatation *value;
+
+  /* search the table with key */
+  if (!shared_model_table) {
+    ml_loge ("The shared model representation is not supported properly!");
+    return NULL;
+  }
+  value = g_hash_table_lookup (shared_model_table, key);
+  if (!value) {
+    ml_logi ("There is no value of the key: %s", key);
```
Comment (dongju-chae) [inline gst/nnstreamer/tensor_filter/tensor_filter_common.c:3000]: maybe `ml_loge`?

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 2]
```diff
@@ -969,6 +986,34 @@ TFLiteCore::setInputTensorDim (const GstTensorsInfo *info)
...
+{
+  int err;
+  TFLiteInterpreter * interpreter_temp = interpreter;
+  if (!gst_tensors_info_is_equal (interpreter->getInputTensorsInfo (),
+          new_interpreter->getInputTensorsInfo ())
+      || !gst_tensors_info_is_equal (interpreter->getOutputTensorsInfo (),
+             new_interpreter->getOutputTensorsInfo ())) {
+    ml_loge ("The model has unmatched tensors info\n");
+    err = -EINVAL;
+    return err;
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:1006]: You may just use `return -EINVAL` and remove `int err;`

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 3]
```diff
@@ -969,6 +986,34 @@ TFLiteCore::setInputTensorDim (const GstTensorsInfo *info)
...
+  }
+
+  interpreter_temp->lock ();
+  interpreter = new_interpreter;
+  interpreter_temp->unlock ();
+
+  interpreter_temp->unref ();
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:1013]: Is it thread-safe? `ref_count` definitely needs a lock or atomic operations.

File: gst/nnstreamer/tensor_filter/tensor_filter_common.c (hunk 1 above)
Comment (kimjh12) [reply gst/nnstreamer/tensor_filter/tensor_filter_common.c]: Fixed!

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc (hunk 2 above)
Comment (kimjh12) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc]: Fixed. Thanks

File: gst/nnstreamer/tensor_filter/tensor_filter_common.c (hunk 1 above)
Comment (kimjh12) [reply gst/nnstreamer/tensor_filter/tensor_filter_common.c]: It's allowed to check value existence.

File: gst/nnstreamer/tensor_filter/tensor_filter_common.c [hunk 4]
```diff
@@ -2786,3 +2801,168 @@ gst_tensor_filter_check_hw_availability (const gchar * name, const accl_hw hw)
...
+ * @param[in] free_callback The callback function to destroy the old interpreter.
+ */
+void
+nnstreamer_filter_shared_model_replace (void *instance, const char *key,
+    void *new_interpreter, void (*replace_callback) (void *, void *),
+    void (*free_callback) (void *))
+{
+  GstTensorFilterSharedModelRepresenatation *model_rep;
+  GList *itr;
+
```
Comment (anyj0527) [inline gst/nnstreamer/tensor_filter/tensor_filter_common.c:2946]: Need to add `UNUSED (instance)` ?

Comment (kimjh12) [reply gst/nnstreamer/tensor_filter/tensor_filter_common.c:2946]: many thanks

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 5]
```diff
@@ -982,6 +985,41 @@ TFLiteCore::setInputTensorDim (const GstTensorsInfo *info)
...
+  TFLiteInterpreter * interpreter_temp = interpreter;
+  if (!gst_tensors_info_is_equal (interpreter->getInputTensorsInfo (),
+          new_interpreter->getInputTensorsInfo ())
+      || !gst_tensors_info_is_equal (interpreter->getOutputTensorsInfo (),
+             new_interpreter->getOutputTensorsInfo ())) {
+    ml_loge ("The model has unmatched tensors info\n");
+    return -EINVAL;
+  }
+
+  interpreter_temp->lock ();
```
Comment (zhoonit) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:1006]: The owner of the class should lock instead? Looks good to go  for now though

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 6]
```diff
@@ -982,6 +985,41 @@ TFLiteCore::setInputTensorDim (const GstTensorsInfo *info)
...
+  interpreter = new_interpreter;
+  interpreter_temp->unlock ();
+
+  return 0;
+}
+
+/**
+ * @brief callback method to replace interpreter for shared model
+ */
+void replace_interpreter (void * instance, void * interperter) {
```
Comment (zhoonit) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:1016]: maybe  `replaceInterpreter` but can be dealt later.

File: gst/nnstreamer/tensor_filter/tensor_filter_common.c [hunk 7]
```diff
@@ -2928,3 +2928,45 @@ nnstreamer_filter_shared_model_remove (void *instance, const char *key,
...
+  if (!shared_model_table) {
+    ml_loge ("The shared model representation is not supported properly!");
+    return;
+  }
+  if (!key) {
+    ml_loge ("The key should NOT be NULL!");
+    return;
+  }
+
+  g_mutex_lock (&mutex);
```
Comment (zhoonit) [inline gst/nnstreamer/tensor_filter/tensor_filter_common.c:2960]: as `replace_callback` is only used here, it seems it's protected and looks valid.

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 8]
```diff
@@ -1101,6 +1104,41 @@ TFLiteCore::setInputTensorDim (const GstTensorsInfo *info)
...
+          new_interpreter->getInputTensorsInfo ())
+      || !gst_tensors_info_is_equal (interpreter->getOutputTensorsInfo (),
+             new_interpreter->getOutputTensorsInfo ())) {
+    ml_loge ("The model has unmatched tensors info\n");
+    return -EINVAL;
+  }
+
+  interpreter_temp->lock ();
+  interpreter = new_interpreter;
+  interpreter_temp->unlock ();
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:1127]: memory leak of ```interpreter_temp``` ?

**Review by myungjoo — APPROVED** [review]
LGTM

**Comment by myungjoo** [comment]
PTAL: @anyj0527

**Review by songgot — APPROVED** [review]
Looks good to me

**Pull request merged**